app/models/assignment.rb

```diff
@@ -379,42 +379,125 @@ def clone_groupings_from(assignment_id)
     group = Group.where(group_name: row.first).first
-    if group.nil?
-      group = Group.new
-      group.group_name = row[0]
+
+    unless group.nil?
+      if group.repo_name != row[1]
+        # CASE: Group already exits but the repo name is different
+        duplicate_group_error = I18n.t('csv.group_with_different_repo',
+                                       group_name: row[0])
+        return duplicate_group_error
+      else
+        any_grouping = Grouping.find_by group_id: group.id
+        if any_grouping.nil?
+          # CASE: Group exists with same repo name but has no grouping
+          #       associated with it for any assignment
+          # Use existing group and create a new grouping between the existing
+          # group and the given students and return without error
+          add_new_grouping_for_group(row, group)
+          return
+        else
+          grouping_for_current_assignment = group.grouping_for_assignment(id)
+          if grouping_for_current_assignment.nil?
+            if same_membership_as_csv_row?(row, any_grouping)
+              # CASE: Group already exists with the same repo name and has a
+              #     grouping for another assignment with the same membership
+              # Use existing group and create a new grouping between the
+              # existing  group and the given students and return without error
+              add_new_grouping_for_group(row, group)
+              return
+            else
+              # CASE: Group already exists with the same repo name and has
+              #     a grouping for another assignment BUT with different
+              #     membership
+              # The existing groupings and the current group is not compatible
+              # Return an error.
+              duplicate_group_error = I18n.t(
+                'csv.group_with_different_membership_different_assignment',
+                group_name: row[0])
+              return duplicate_group_error
+            end
+          else
+            if same_membership_as_csv_row?(row,
+                                           grouping_for_current_assignment)
+              # CASE: Group already exists with the same repo name and also has
+              #     a grouping for the current assignment with the same
+              #     membership
+              # No new group or grouping created. Since the exact group given by
+              # the csv file already exists treat this as a successful case
+              # and don't return an error
+              return
+            else
+              # CASE: Group already exists with the same repo name and has a
+              #     grouping for the current assignment BUT the membership is
+              #     different.
+              # Return error since the membership is different
+              duplicate_group_error = I18n.t(
+                'csv.group_with_different_membership_current_assignment',
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.